### ISSUE TITLE
Access modifier pub mut allows modification outside parent module

### ISSUE THREAD

**Issue by Ekopalypse** [vlang/v#9894]
V full version: V 0.2.2 bca9409.3877522
OS: windows, Microsoft Windows 7 Professional v7601 64-bit

**What did you do?**
Created a file foo.v
~~~v
module main
import bar

fn main() {
	mut x := bar.Bar{}
	x.i += 1
	println(x.i)
	x.test()
	println(x.i)
}
~~~

in addition a directory bar with a file bar.v

~~~v
module bar

pub struct Bar {
pub mut:
	i int
}

pub fn (mut b Bar) test() {
	b.i++
}
~~~

**What did you expect to see?**
Compiler error reporting that field "i" is immutable
 
**What did you see instead?**
No error, "i" could be changed.

**Comment by mcastorina** [comment]
This looks like correct behavior to me. Why do you expect `i` to be immutable when it is declared as `mut`?

**Comment by Larpon** [comment]
`pub`lic `mut`able. This isn't a bug - it's a feature 🙂

**Comment by Gladear** [comment]
Reopening as, whether this is or isn't the behavior expected, it doesn't fit the [documentation].(https://github.com/vlang/v/blob/master/doc/docs.md#access-modifiers)
```v
pub mut:
	e int // public, but mutable only in parent module
```

**Comment by medvednikov** [comment]
Yes, it's been like this forever, and always bugging me, because I just don't know what syntax to use for **really** public fields.

The original idea was `pub mut mut:`, but it's ugly. `__global:` feels too dangerous.

**Comment by JalonSolov** [comment]
If you don't declare it `pub`, it is still accessible anywhere within the same module.  I think it is working as it should already, and the documentation just needs to be adjusted.

If you really, truly need to mark something as private to only the file it is declared in, that will require a new keyword.

**Comment by mcastorina** [comment]
What are **really** public fields? I thought the idea of fields were either they can be accessed from outside the module or not.

**Comment by JalonSolov** [comment]
That's what pub means.  I think the question is whether or not adding `mut` to `pub` vars should make them modifiable outside the module, or whether some other syntax should be used for that.

I vote for leaving things as they are now, but add a [local_mut] or [module_mut] attribute if you only want them modifiable inside the module, but still `pub` outside.

**Comment by ken0x0a** [comment]
This issue, perhaps, should be moved to discussion.

Anyway,

Is it too confusing, to distinguish `pub mut:` and `mut pub:`?
- `pub`licly `mut`able; and
- `mut`able (locally, and) `pub`licly (readable).

I don't know about English though,
different order should mean different (should have only one way right?).

Just using current keyword looks simpler and clean.
More keyword makes language, looks more complicated.


```v
struct AAA {
pub mut:
	public_mutable string
mut pub:
	local_mutable string
}
```

**Comment by dumblob** [comment]
Distinguishing `pub mut` and `mut pub` is not acceptable. We need something **much** clearer - perhaps what @JalonSolov suggested.

**Comment by ken0x0a** [comment]
"use the wrong order" means compiler will let us know what's wrong.


> The original idea was `pub mut mut:`, but it's ugly. `__global:` feels too dangerous.

from the idea `pub mut mut:`, `mut pub:` can be understood as not mutable in public.

The users including me think, this is not clear and/or bad design, but after try using it, this can be really good.

`pub mut:` and `mut pub:` is _4_ characters different, on the other hand `var` and `&var` is only _1_ character added,
so I guess only few people claim about this after getting used to it.


And also,
if there is no difference between them, one of them should be wrong syntax.
Otherwise, 2 ways force users to decide which to use.
Should be one way, to avoid useless discussion, can be happen in future.

**Comment by JalonSolov** [comment]
No, using the wrong order means you could put `mut pub` when you meant `pub mut`, and there is no way for the compiler to tell you that you did it in the wrong order.

**Comment by ken0x0a** [comment]
> No, using the wrong order means you could put `mut pub` when you meant `pub mut`, and there is no way for the compiler to tell you that you did it in the wrong order.

That's miscommunication.

I know the current behavior.
I mean, compiler should show error, and that need to be implemented.

**Comment by JalonSolov** [comment]
What type of error is it supposed to show?  How does the compiler know that you meant `pub mut` when it sees `mut pub`, or vice versa?

**Comment by ken0x0a** [comment]
> What type of error is it supposed to show? How does the compiler know that you meant `pub mut` when it sees `mut pub`, or vice versa?

If what `pub mut` means & what `mut pub` means are same ("publicly mutable" as you posted above), then it can something like "`mut pub` is not allowed, use `pub mut` instead" or,... whatever understantable.
If we need to do that, I think it's easier to do in `v.parser`.

But, I expect those to have different meaning, and used differently.
I prefer `fn` over `function`.
I prefer `pub` over `export`.
I prefer `mut pub` over `pub local_mut` and `local_mut pub` (I'm sure that having 2 syntaxes for often used case is not cool).

I appreciate that Vlang doesn't have `function` like to verbose syntax at often used place.

`mut pub` is easy enough to understand because it will be often used, and also it's way easier to notice the difference when I write.

```v
struct Bcd {
mut:
	hash string
mut pub:
	local_mutable string
pub:
	public_immutable int
pub mut:
	public_mutable string
}
```

**Comment by mcastorina** [comment]
I think we need to take a step back and look at the system as a whole. In my opinion, all struct members should be mutable by default and we can get rid of the `mut` keyword in the struct context. Immutability should be handled at the instance level. If a piece of code shouldn't modify the struct, use an immutable reference.

```v
module foo

pub struct Foo {
    bar int
pub:
    baz int
}

// Foo{} allowed here because this module has access to all struct members
pub fn new() Foo { return Foo{} }
```

```v
module main

mut x := foo.Foo{} // not allowed because not all struct members are public

mut x := foo.new()
println(x.bar)  // not allowed, bar is not public
x.bar = 0       // not allowed, bar is not public
println(x.baz)  // allowed, baz is public
x.baz = 0       // allowed, baz is public and x is mutable

y := foo.new()
y.baz = 0       // not allowed, y is immutable
```

This is how Rust does it, and I think they got it right.

**Comment by felipensp** [comment]
Something to do on this issue yet @medvednikov ?

**Comment by Lartrax** [comment]
What about `vis` and `vis mut`? `vis` meaning visible outside of module.
`pub` would then work the same as now, meaning public access. While `vis` would only grant visibility outside the module.

Maybe singular `vis` and `pub` could have different behavior, though I don't know what difference it would be. Or we only add `vis mut`.

```v
pub struct App {
	internal int
mut:
	internal_mut int
pub:
	public int      // visible outside module, immutable
pub mut:
	public_mut int  // visible and mutable outside of module
vis:
	visible int     // visible outside module (identical to pub) 
vis mut:
	visible_mut int // visible outside module, but only mutable internally
}
```

**Comment by medvednikov** [comment]
How about `@readonly` or `@ro` attribute for `pub mut` fields for "external read, internal mut"?

**Comment by JalonSolov** [comment]
`@[local]`, or `@[module_mut]`, perhaps?

**Comment by GGRei** [comment]
I also like JalonSolov suggestion: `module_mut`.

It names the rule directly the field is readable publicly, but mutable only from the defining module.

Example:

```v
module bar

pub struct Bar {
@[module_mut]
pub:
    i int
}

pub fn (mut b Bar) inc() {
    b.i++
}
```

Then from another module:

```v
mut x := bar.Bar{}
println(x.i) // OK, public read
x.i++        // error: field `i` is mutable only inside module `bar`
x.inc()      // OK
```

So the meaning would be:

- `pub:` = public read, not externally mutable
- `pub mut:` = public read/write, current behavior
- `@[module_mut] pub:` = public read, mutable only inside the defining module

This avoids changing the existing meaning of `pub mut:`, so it should be less breaking, while still matching the documented/useful case of “publicly readable, internally mutable”.

Of course, this is only my opinion. 

I would also suggest introducing this as a V2-only feature, since V1 is going away.

**Comment by medvednikov** [comment]
fixed in https://github.com/vlang/v/pull/27268 once it's merged
